web/client-api/src/main/java/io/deephaven/web/client/api/JsPartitionedTable.java

```diff
@@ -83,12 +88,18 @@ public Promise<JsPartitionedTable> refetch() {
             descriptor = PartitionedTableDescriptor.deserializeBinary(w.getDataAsU8());
 
             keyColumnTypes = new ArrayList<>();
+            InitialTableDefinition tableDefinition = WebBarrageUtils.readTableDefinition(
+                    WebBarrageUtils.readSchemaMessage(descriptor.getConstituentDefinitionSchema_asU8()));
             ColumnDefinition[] columnDefinitions = WebBarrageUtils.readColumnDefinitions(
                     WebBarrageUtils.readSchemaMessage(descriptor.getConstituentDefinitionSchema_asU8()));
-            for (int i = 0; i < columnDefinitions.length; i++) {
-                ColumnDefinition columnDefinition = columnDefinitions[i];
+            columns = new Column[0];
+            keyColumns = new Column[0];
+            for (ColumnDefinition columnDefinition : columnDefinitions) {
+                Column column = columnDefinition.makeJsColumn(columns.length, tableDefinition.getColumnsByName());
+                columns[columns.length] = column;
                 if (descriptor.getKeyColumnNamesList().indexOf(columnDefinition.getName()) != -1) {
                     keyColumnTypes.add(columnDefinition.getType());
+                    keyColumns[keyColumns.length] = column;
                 }
             }
```

Comment:
After all of the items are added to the new arrays, freeze them with `JsObject.freeze(...)`, so they are safe to return to the user.